tests/test_aggregates.py

```diff
@@ -0,0 +1,270 @@
+        call_params = ['str', 1, None]
+        for param in call_params:
+            with pytest.raises(TypeError):
+                aggregator()(param)
+
+    def test_call_without_decorator(self):
+        aggregate_instance = aggregator()
+        with pytest.raises(TypeError):
+            aggregate_instance()
+
+    def test_call_with_decorator(self):
+        @aggregator()
+        def test_function(x):
+            return x
+
+        assert hasattr(test_function, '_flow_aggregate')
+
+    def test_groups_of_invalid_num(self):
+        invalid_values = [{}, 'str', -1, -1.5]
+        for invalid_value in invalid_values:
+            with pytest.raises((TypeError, ValueError)):
+                aggregator.groupsof(invalid_value)
+
+    def test_group_by_invalid_key(self):
+        with pytest.raises(TypeError):
+            aggregator.groupby(1)
+
+    def test_groupby_with_valid_type_default_for_Iterable(self):
+        aggregator.groupby(['half', 'even'], default=[-1, -1])
+
+    def test_groupby_with_invalid_type_default_key_for_Iterable(self):
+        with pytest.raises(TypeError):
+            aggregator.groupby(['half', 'even'], default=-1)
+
+    def test_groupby_with_invalid_length_default_key_for_Iterable(self):
+        with pytest.raises(ValueError):
+            aggregator.groupby(['half', 'even'], default=[-1, -1, -1])
+
+    def test_aggregate_hashing(self):
+        # Since we need to store groups on a per aggregate basis in the project,
+        # we need to be sure that the aggregates are hashing and compared correctly.
+        # This test ensures this feature.
+
+        def helper_default_aggregator_function(jobs):
+            return [jobs]
+
+        def helper_non_default_aggregator_function(jobs):
+            for job in jobs:
+                yield (job,)
+
+        list_of_aggregates = [
+            aggregator(), aggregator(), aggregator(helper_default_aggregator_function),
+            aggregator(helper_non_default_aggregator_function),
+            aggregator(helper_non_default_aggregator_function),
+            aggregator.groupsof(1), aggregator.groupsof(1),
+            aggregator.groupsof(2), aggregator.groupsof(3), aggregator.groupsof(4),
+            aggregator.groupby('half'), aggregator.groupby('half'), aggregator.groupby('even'),
+            aggregator.groupby('half', -1), aggregator.groupby('even', -1),
+            aggregator.groupby(['half', 'even'], default=[-1, -1])
+        ]
+
+        # The above list contains 11 distinct aggregator objects and some duplicates.
+        # When this list is converted to set, then these objects hashed first and
+        # then compared. Since sets don't carry duplicate values, we test whether the
+        # length of the set obtained from the list is equal to 11 or not.
+        assert len(set(list_of_aggregates)) == 11
+
+
+# Test the _AggregatesStore and _DefaultAggregateStore classes
+class TestAggregateStoring(AggregateProjectSetup):
```

Comment:
Can we test the `__hash__` and `__eq__` of this class given the typo I found.

Reply:
Yes, I think I forgot to write those. Sorry about that. Will do it.

Reply:
Have these tests been written?

Reply:
@vyasr yes, I have written the tests.
You can have a look at them (They are the last 4 tests of this class)